libminifi/src/core/Processor.cpp

```diff
@@ -268,6 +268,7 @@ void Processor::onTrigger(const std::shared_ptr<ProcessContext> &context, const
 
 bool Processor::isWorkAvailable() {
   // We have work if any incoming connection has work
+  std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
Accessing incoming connections
Not sure it's needed, but won't hurt, it's done the same way in "flowFilesQueued()" and "flowFileOutGoingFull()" above in this file, also the functions adding/removing connections lock this mutex as well.

Reply:
We should definitely think about introducing generic thread-safe proxy containers. (Whereby you can only access the encapsulated object through a lock handle.)

Reply:
+1 to that. Definitely not in scope of this PR, but I strongly support.

Reply:
Agreed.

Reply:
One of my hopes was that we could make some asks to Folly to use their thread pool(s)...I believe they've already filled those gaps and I think they have some useful containers. It would be a massive undertaking but Folly has some nice utilities that are well tested.

Reply:
> My guess would be that we need to use `Connectable::work_available_mutex_` instead of `Processor::mutex_`.
> 
> I find it a very bad design decision to expose threading and signaling primitives in `Connectable`. `isWorkAvailable()` should be a source object (of "work") that can be checked for readiness and waited for. That would break an important API, though, so just leaving it here.

In theory you are most probably right, but if you look at that file, all the other functions that touch incoming and outgoing connections lock this mutex. 
I'm fine with a follow-up ticket to clean this up, but this PR wasn't meant to do that.